mitiq/lre/multivariate_scaling/layerwise_folding.py

```diff
@@ -0,0 +1,195 @@
+
+        Args:
+            input_circuit: Checks whether this circuit is able to be folded.
+
+        Raises:
+            UnfoldableCircuitError:
+                * If the circuit has intermediate measurements.
+                * If the circuit has non-unitary channels which are not
+                    terminal measurements.
+
+        Returns:
+            num_layers: the number of layers in the input circuit without the
+                terminal measurements.
+    """
+    circuit = deepcopy(input_circuit)
+    _check_foldable(circuit)
+    _pop_measurements(circuit)
+    num_layers = len(circuit)
+    return num_layers
+
+
+def _get_chunks(
+    input_circuit: Circuit, num_chunks: Optional[int] = None
+) -> List[Circuit]:
+    """Splits a circuit into approximately equal chunks.
+
+    Adapted from:
+    https://stackoverflow.com/questions/2130016/splitting-a-list-into-n-parts-of-approximately-equal-length
+
+        Args:
+            input_circuit: Circuit of interest.
+            num_chunks: Number of desired approximately equal chunks
+                * when num_chunks == num_layers, the original circuit is
+                    returned
+                * when num_chunks == 1, the entire circuit is chunked into 1
+                    layer
+
+
+        Returns:
+            split_circuit: Circuit of interest split into approximately equal
+                chunks
+    """
+    num_layers = _get_num_layers_without_measurements(input_circuit)
+    if num_chunks is None:
+        num_chunks = num_layers
+
+    if num_chunks == 0:
+        raise ValueError("The number of chunks should be >= to 1.")
+
+    if num_chunks > num_layers:
+        raise ValueError(
+            "Number of chunks > the number of layers in the circuit."
+        )
+
+    k, m = divmod(num_layers, num_chunks)
+    split_circuit = (
+        input_circuit[i * k + min(i, m) : (i + 1) * k + min(i + 1, m)]
+        for i in range(num_chunks)
+    )
+    return list(split_circuit)
+
+
+def _get_scale_factor_vectors(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> List[Tuple[Any, ...]]:
+    """Returns the patterned scale factor vectors required for multivariate
+    extrapolation.
+
+        Args:
+            input_circuit: Circuit to be scaled
+            degree: Degree of the multivariate polynomial
+            fold_multiplier: Scaling gap required by unitary folding
+
+        Returns:
+            scale_factor_vectors: Multiple variations of scale factors for each
+                layer in the input circuit
+    """
+
+    circuit_chunks = _get_chunks(input_circuit, num_chunks)
+    num_layers = len(circuit_chunks)
+
+    # find the exponents of all the monomial terms required for the folding
+    # pattern
+    pattern_full = []
+    for i in range(degree + 1):
+        for j in itertools.combinations_with_replacement(range(num_layers), i):
+            pattern = [0] * num_layers
```

Comment:
Instead of defining this as a list and then converting it to an `np.array`, you can just define it as a `np.array` of zeros

```py
pattern = np.zeros(num_layers)
```

then

```py
pattern_full.append(tuple(fold_multiplier * pattern))
```